core/src/com/unciv/logic/city/IConstruction.kt

```diff
@@ -1,23 +1,99 @@
 package com.unciv.logic.city
 
 import com.unciv.logic.civilization.CivilizationInfo
+import com.unciv.models.ruleset.Unique
 import com.unciv.models.stats.INamed
+import com.unciv.models.stats.Stat
 import com.unciv.ui.utils.Fonts
+import kotlin.math.pow
 import kotlin.math.roundToInt
 
 interface IConstruction : INamed {
-    fun getProductionCost(civInfo: CivilizationInfo): Int
-    fun getGoldCost(civInfo: CivilizationInfo): Int
     fun isBuildable(cityConstructions: CityConstructions): Boolean
     fun shouldBeDisplayed(cityConstructions: CityConstructions): Boolean
     fun postBuildEvent(cityConstructions: CityConstructions, wasBought: Boolean = false): Boolean  // Yes I'm hilarious.
     fun getResourceRequirements(): HashMap<String,Int>
-    fun canBePurchased(): Boolean
+}
+
+interface INonPerpetualConstruction : IConstruction, INamed {
+    val hurryCostModifier: Int
+    val uniqueObjects: List<Unique>
+    val uniques: List<String>
+
+    fun getProductionCost(civInfo: CivilizationInfo): Int
+    fun getStatBuyCost(cityInfo: CityInfo, stat: Stat): Int?
+    
+    private fun getMatchingUniques(uniqueTemplate: String): Sequence<Unique> {
+        return uniqueObjects.asSequence().filter { it.placeholderText == uniqueTemplate }
+    }
+    
+    fun canBePurchasedWithStat(cityInfo: CityInfo, stat: Stat): Boolean {
+        if (stat in listOf(Stat.Production, Stat.Happiness)) return false
+        if ("Cannot be purchased" in uniques) return false
+        if (stat == Stat.Gold) return !uniques.contains("Unbuildable")
+        if (getMatchingUniques("Can be purchased with []")
+                .filter { it.params[0] == stat.name }
+                .any()
+        ) return true
+        if (getMatchingUniques("Can be purchased with [] []")
+                .filter { it.params[0] == stat.name && cityInfo.matchesFilter(it.params[1]) }
+                .any()
+        ) return true
+        if (getMatchingUniques("Can be purchased for [] []")
+                .filter { it.params[1] == stat.name}
+                .any()
+        ) return true
+        if (getMatchingUniques("Can be purchased for [] [] []")
+                .filter { it.params[1] == stat.name && cityInfo.matchesFilter(it.params[2]) }
+                .any()
+        ) return true
```

Comment:
instead of .filter{...}.any() you can just .any{...}
Also, it's hard to tell at a glance what each parameter is, so please give examples / decorate with e.g.
// "Can be purchased for [amount] [statName] [cityFilter]"

Reply:
ALSO
we can just reduce these to 2 uniques - "Can be purchased for [amount] [statName] [cityFilter]", "Can be purchased with [statName] [cityFilter]"
If it's in all cities then cityFilter will be "in all cities"
Much easier than maintaining 4 uniques for this